benchmarks/Microsoft.AspNetCore.Routing.Performance/LinkGeneration/SingleRouteWithConstraintsBenchmark.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using BenchmarkDotNet.Attributes;
+using Microsoft.AspNetCore.Routing.Tree;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Routing.LinkGeneration
+{
+    public class SingleRouteWithConstraintsBenchmark : LinkGenerationBenchmarkBase
+    {
+        private TreeRouter _treeRouter;
+        private LinkGenerator _linkGenerator;
+        private RequestContext _requestContext;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            var template = "Customers/Details/{category}/{region}/{id:int}";
+            var defaults = new { controller = "Customers", action = "Details" };
+            var requiredValues = new { controller = "Customers", action = "Details" };
+
+            // Endpoint routing related
+            var endpoint = CreateEndpoint(template, defaults, requiredValues: requiredValues);
+            var services = CreateServices(endpoint);
+            _linkGenerator = services.GetRequiredService<LinkGenerator>();
+
+            // Attribute routing related
+            var treeRouteBuilder = services.GetRequiredService<TreeRouteBuilder>();
+            CreateOutboundRouteEntry(treeRouteBuilder, endpoint);
+            _treeRouter = treeRouteBuilder.Build();
+
+            _requestContext = CreateCurrentRequestContext();
+        }
+
+        [Benchmark(Baseline = true)]
+        public void TreeRouter()
+        {
+            var virtualPathData = _treeRouter.GetVirtualPath(new VirtualPathContext(
+                _requestContext.HttpContext,
+                ambientValues: _requestContext.AmbientValues,
+                values: new RouteValueDictionary(
```

Comment:
IMO initialize the RVD once in Setup and reuse it. It reduces some of the allocation noise in the results.
Could do this for all benchmarks.

Reply:
I agree with Kiran here. Typically a user is passing a new RVD into each callsite.